ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c

```diff
@@ -6905,39 +6905,51 @@ bdb_public_private_open(backend *be, const char *db_filename, int rw, dbi_env_t
     int flags;
     int rc;
 
-    /* Either filename is an existing regular file
-     *  or the "home" directory where txn logs are
-     */
+    slapi_ch_free_string(&conf->bdb_dbhome_directory);
+    if (li->li_directory == NULL) {
+        /* Either filename is an existing regular file
+         *  or the "home" directory where txn logs are
+         */
 
-    PL_strncpyz(dbhome, db_filename, MAXPATHLEN);
-    if (stat(dbhome, &st) == 0) {
-        if (S_ISDIR(st.st_mode)) {
-            li->li_directory = slapi_ch_strdup(dbhome);
-        } else if (S_ISREG(st.st_mode)) {
+        PL_strncpyz(dbhome, db_filename, MAXPATHLEN);
+        if (stat(dbhome, &st) == 0) {
+            if (S_ISDIR(st.st_mode)) {
+                li->li_directory = slapi_ch_strdup(dbhome);
+            } else if (S_ISREG(st.st_mode)) {
+                getdir(dbhome, NULL);
+                li->li_directory = slapi_ch_strdup(db_filename);
```

Comment:
The patch does not change that code but I do no understand this line: li_directory refers to a file not a directory ?
Does it work ?

Reply:
li->li_directory is supposed to be the instance directory (i.e .../db/userroot )
bdb_public_private_open is only called by dbscan to open the database.
And the db_filename is the instance path (i.e: .../db/userroot/cn.db) or it could be the global database (i.e .../db when trying to list all the existing instances)

 The puzzling line is  handling an error case (the provided path is wrong) as we cannot determine the li_directory, we provides directly the path (that will  generates an errors later on with a reasonable error message )
 Does it works: ?  No but that is expected. and it provides a reasonable error ( Not a directory) :
 
dbscan -D bdb -f $DB/userRoot/cn.db/foo.db 
bdb_public_private_open failed with db error 20 : Not a directory
Can't initialize db plugin: bdb